reviewer/reviewer.go

```diff
@@ -0,0 +1,163 @@
+			if c.Status == common.StatusPending {
+				r = append(r, findLeafChildren(*c)...)
+			}
+		}
+	}
+	return r
+}
+
+// select n random values from the list of users without reuse
+func selectRandomUsers(n int, users []string, r *rand.Rand) []string {
+	selected := make(map[int]bool)
+
+	var selections []string
+	if n == 0 {
+		return selections
+	}
+	if n >= len(users) {
+		return users
+	}
+
+	for i := 0; i < n; i++ {
+		for {
+			i := r.Intn(len(users))
+			if !selected[i] {
+				selected[i] = true
+				selections = append(selections, users[i])
+				break
+			}
+		}
+	}
+	return selections
+}
+
+func shoveIntoMap(m map[string]struct{}, u []string) map[string]struct{} {
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+	return m
+}
+
+func selectTeamMembers(prctx pull.Context, allTeams []string, r *rand.Rand) ([]string, error) {
+	randomTeam := allTeams[r.Intn(len(allTeams))]
+	teamInfo := strings.Split(randomTeam, "/")
+	teamMembers, err := prctx.ListTeamMembers(teamInfo[0], teamInfo[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+	}
+	return teamMembers, nil
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, r *rand.Rand) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		allUsers = shoveIntoMap(allUsers, child.Rule.RequestedUsers)
+
+		if len(child.Rule.RequestedTeams) > 0 {
+			teamMembers, err := selectTeamMembers(prctx, child.Rule.RequestedTeams, r)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for teams, skipping team member selection")
+			}
+			allUsers = shoveIntoMap(allUsers, teamMembers)
+		}
+
+		if len(child.Rule.RequestedOrganizations) > 0 {
+			randomOrg := child.Rule.RequestedOrganizations[r.Intn(len(child.Rule.RequestedOrganizations))]
+			orgMembers, err := prctx.ListOrganizationMembers(randomOrg)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for org %s, skipping org member selection", randomOrg)
+			}
+			allUsers = shoveIntoMap(allUsers, orgMembers)
+		}
+
+		allCollaboratorPermissions, err := prctx.ListRepositoryCollaborators()
+		if err != nil {
+			return nil, errors.Wrap(err, "Unable to list repository collaborators")
+		}
+
+		if child.Rule.RequestedAdmins {
+			var repoAdmins []string
+			for _, c := range allCollaboratorPermissions {
+				if allCollaboratorPermissions[c] == "admin" {
+					repoAdmins = append(repoAdmins, c)
+				}
+			}
+			allUsers = shoveIntoMap(allUsers, repoAdmins)
+		}
+
+		if child.Rule.RequestedWriteCollaborators {
+			var repoCollaborators []string
+			for _, c := range allCollaboratorPermissions {
+				if allCollaboratorPermissions[c] == "write" {
```

Comment:
Does this show up as `write` or `push` in the API responses? I know we've run into this confusion before when _setting_ permissions.

Better yet, I saw the tests use a constant defined in the `common` package; let's use that here too.

Reply:
They show up as `admin`, `write`: https://developer.github.com/v4/enum/repositorypermission/

You're right I should just use the common constants!